drivers/counter/rtc_mcp7940n.c

```diff
@@ -0,0 +1,785 @@
+		if (rc < 0) {
+			goto out;
+		}
+		interrupt_pending |= (1 << ALARM1_ID);
+	}
+
+out:
+	k_sem_give(&data->lock);
+
+	if (rc == 0) {
+		rc = (int)interrupt_pending;
+	}
+	return rc;
+}
+
+static int mcp7940n_counter_get_top_value(const struct device *dev)
+{
+	return UINT32_MAX;
+}
+
+static int mcp7940n_init(const struct device *dev)
+{
+	struct mcp7940n_data *data = dev->data;
+	const struct mcp7940n_config *cfg = dev->config;
+	int rc = 0;
+	time_t unix_time = 0;
+
+	/* Initialize and take the lock */
+	k_sem_init(&data->lock, 0, 1);
+
+	if (!device_is_ready(cfg->i2c_dev)) {
+		LOG_ERR("I2C device not found");
+		rc = -EINVAL;
+		goto out;
+	}
+
+	rc = read_time(dev, &unix_time);
+	if (rc < 0) {
+		goto out;
+	}
+
+	rc = set_day_of_week(dev, &unix_time);
+	if (rc < 0) {
+		goto out;
+	}
+
+	/* Set 24-hour time */
+	data->registers.rtc_hours.twelve_hr = false;
+	rc = write_register(dev, REG_RTC_HOUR, data->registers.rtc_hours);
+	if (rc < 0) {
+		goto out;
+	}
+
+	/* Configure alarm interrupt gpio */
+	if (cfg->int_gpios.ctrl != NULL) {
+		data->gpio_int = device_get_binding(cfg->int_gpios.ctrl);
+
+		if (!data->gpio_int) {
+			rc = -ENOTSUP;
+			LOG_ERR("GPIO interrupt not found");
+		} else {
+			data->mcp7940n = dev;
+			k_work_init(&data->alarm_work, mcp7940n_work_handler);
+
+			gpio_pin_configure(data->gpio_int, cfg->int_gpios.pin,
+					   GPIO_INPUT | cfg->int_gpios.flags);
+			gpio_pin_interrupt_configure(data->gpio_int, cfg->int_gpios.pin,
+					   GPIO_INT_EDGE_TO_ACTIVE);
+
+			gpio_init_callback(&data->int_callback, mcp7940n_init_cb,
+					   BIT(cfg->int_gpios.pin));
+
+			gpio_add_callback(data->gpio_int, &data->int_callback);
+
+			/* Configure interrupt polarity */
+			if ((cfg->int_gpios.flags & GPIO_ACTIVE_LOW) == GPIO_ACTIVE_LOW) {
+				data->int_active_high = false;
+			} else {
+				data->int_active_high = true;
+			}
+			data->alm0_registers.alm_weekday.alm_pol = data->int_active_high;
+			data->alm1_registers.alm_weekday.alm_pol = data->int_active_high;
+			rc = write_register(dev, REG_ALM0_WDAY,
+					    data->alm0_registers.alm_weekday);
+			rc = write_register(dev, REG_ALM1_WDAY,
+					    data->alm1_registers.alm_weekday);
+		}
+	}
+out:
+	k_sem_give(&data->lock);
+
+	return rc;
+}
+
+static const struct counter_driver_api mcp7940n_api = {
+	.start = mcp7940n_counter_start,
+	.stop = mcp7940n_counter_stop,
+	.get_value = mcp7940n_counter_get_value,
+	.set_alarm = mcp7940n_counter_set_alarm,
+	.cancel_alarm = mcp7940n_counter_cancel_alarm,
+	.set_top_value = mcp7940n_counter_set_top_value,
+	.get_pending_int = mcp7940n_counter_get_pending_int,
+	.get_top_value = mcp7940n_counter_get_top_value,
+};
+
+static const struct mcp7940n_config mcp7940n_0_config = {
+	.generic = {
+		.max_top_value = UINT32_MAX,
+		.freq = 1,
+		.flags = COUNTER_CONFIG_INFO_COUNT_UP,
+		.channels = 2,
+	},
+	.i2c_dev = DEVICE_DT_GET(DT_INST_BUS(0)),
+	.addr = DT_INST_REG_ADDR(0),
+#if DT_INST_NODE_HAS_PROP(0, int_gpios)
```

Comment:
By switching to gpio_dt_spec, this can just use GPIO_DT_SPEC_GET_OR; you can grep around the tree for examples.

Reply:
Updated.